protocol/msgpack/unaggregated_iterator.go

```diff
@@ -0,0 +1,373 @@
+}
+
+func (it *unaggregatedIterator) decodeCounter() {
+	numExpectedFields, numActualFields, ok := it.checkNumFieldsForType(counterType)
+	if !ok {
+		return
+	}
+	it.metric.Type = unaggregated.CounterType
+	it.metric.ID = it.decodeID()
+	it.metric.CounterVal = int64(it.decodeVarint())
+	it.skip(numActualFields - numExpectedFields)
+}
+
+func (it *unaggregatedIterator) decodeBatchTimer() {
+	numExpectedFields, numActualFields, ok := it.checkNumFieldsForType(batchTimerType)
+	if !ok {
+		return
+	}
+	it.metric.Type = unaggregated.BatchTimerType
+	it.metric.ID = it.decodeID()
+	numValues := it.decodeArrayLen()
+	values := it.floatsPool.Get(numValues)
+	for i := 0; i < numValues; i++ {
+		values = append(values, it.decodeFloat64())
+	}
+	it.metric.BatchTimerVal = values
+	it.skip(numActualFields - numExpectedFields)
+}
+
+func (it *unaggregatedIterator) decodeGauge() {
+	numExpectedFields, numActualFields, ok := it.checkNumFieldsForType(gaugeType)
+	if !ok {
+		return
+	}
+	it.metric.Type = unaggregated.GaugeType
+	it.metric.ID = it.decodeID()
+	it.metric.GaugeVal = it.decodeFloat64()
+	it.skip(numActualFields - numExpectedFields)
+}
+
+func (it *unaggregatedIterator) decodePolicy() policy.Policy {
+	numExpectedFields, numActualFields, ok := it.checkNumFieldsForType(policyType)
+	if !ok {
+		return policy.Policy{}
+	}
+	resolution := it.decodeResolution()
+	retention := it.decodeRetention()
+	p := policy.Policy{Resolution: resolution, Retention: retention}
+	it.skip(numActualFields - numExpectedFields)
+	return p
+}
+
+func (it *unaggregatedIterator) decodeResolution() policy.Resolution {
+	numActualFields := it.decodeNumObjectFields()
+	resolutionType := it.decodeObjectType()
+	numExpectedFields, numActualFields, ok := it.checkNumFieldsForTypeWithActual(
+		resolutionType,
+		numActualFields,
+	)
+	if !ok {
+		return policy.EmptyResolution
+	}
+	switch resolutionType {
+	case knownResolutionType:
+		resolutionValue := policy.ResolutionValue(it.decodeVarint())
+		it.skip(numActualFields - numExpectedFields)
+		if it.err != nil {
+			return policy.EmptyResolution
+		}
+		resolution, err := resolutionValue.Resolution()
+		it.err = err
+		return resolution
+	case unknownResolutionType:
+		window := time.Duration(it.decodeVarint())
+		precision := xtime.Unit(it.decodeVarint())
+		it.skip(numActualFields - numExpectedFields)
+		return policy.Resolution{Window: window, Precision: precision}
+	default:
+		it.err = fmt.Errorf("unrecognized resolution type %v", resolutionType)
+		return policy.EmptyResolution
+	}
+}
+
+func (it *unaggregatedIterator) decodeRetention() policy.Retention {
+	numActualFields := it.decodeNumObjectFields()
+	retentionType := it.decodeObjectType()
+	numExpectedFields, numActualFields, ok := it.checkNumFieldsForTypeWithActual(
+		retentionType,
+		numActualFields,
+	)
+	if !ok {
+		return policy.EmptyRetention
+	}
+	switch retentionType {
+	case knownRetentionType:
+		retentionValue := policy.RetentionValue(it.decodeVarint())
```

Comment:
Should we also check that this is a known good enum value? i.e. what if the int is out of range?